scripts/test/bs-wrapper.js

```diff
@@ -43,6 +45,49 @@ async function hasRunningBuild() {
 }
 
 function runTests() {
-  const [command, ...args] = process.argv.slice(2)
-  return spawnCommand(command, args)
+  return new Promise((resolve) => {
+    const [command, ...args] = process.argv.slice(2)
+
+    const child = spawn(command, args, {
+      stdio: ['inherit', 'pipe', 'pipe'],
+      env: { ...process.env, FORCE_COLOR: true },
+    })
+
+    let output = ''
+    let isKilled = false
+    let timeoutId
+
+    child.stdout.pipe(process.stdout)
+    child.stdout.on('data', onOutput)
+
+    child.stderr.pipe(process.stderr)
+    child.stderr.on('data', onOutput)
+
+    child.on('exit', (code) => {
+      resolve(!isKilled && code === 0)
+    })
+
+    function onOutput(data) {
+      output += data
+
+      if (hasUnrecoverableFailure(output)) {
+        killIt('unrecoverable failure')
+      }
+
+      clearTimeout(timeoutId)
+      setTimeout(() => killIt('no output timeout'), NO_OUTPUT_TIMEOUT)
+    }
+
+    function killIt(message) {
+      if (!isKilled) {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏ Do we need this `isKilled` variable? Instead, can we return on line 74? (unless `child.kill('SIGTERM')` does not terminate immediately and somehow some more output is processed.

Reply:
> unless child.kill('SIGTERM') does not terminate immediately and somehow some more output is processed

"SIGTERM" is typically a "graceful exit" -> the process can catch it and continue its execution to clean things up before exiting. So it is possible to have some output after a SIGTERM.

In practice, I'm not sure it happens in this particular situation (= karma and wdio processes). Sending the signal multiple times shouldn't hurt, so let's simplify the code!